radix-engine-interface/src/blueprints/resource/fungible/fungible_resource_manager.rs

```diff
@@ -4,21 +4,131 @@ use crate::*;
 use arbitrary::Arbitrary;
 use radix_engine_common::data::manifest::model::ManifestAddressReservation;
 use radix_engine_common::types::*;
+use radix_engine_interface::api::node_modules::auth::ToRoleEntry;
 use radix_engine_interface::api::node_modules::metadata::MetadataInit;
 use radix_engine_interface::api::node_modules::ModuleConfig;
-use sbor::rust::collections::BTreeMap;
 
 pub const FUNGIBLE_RESOURCE_MANAGER_BLUEPRINT: &str = "FungibleResourceManager";
 
 pub const FUNGIBLE_RESOURCE_MANAGER_CREATE_IDENT: &str = "create";
 
+#[cfg_attr(feature = "radix_engine_fuzzing", derive(Arbitrary))]
+#[derive(Default, Debug, Clone, Eq, PartialEq, ScryptoSbor, ManifestSbor)]
+pub struct FungibleResourceFeatures {
+    pub mintable: Option<MintableRoles<RoleDefinition>>,
+    pub burnable: Option<BurnableRoles<RoleDefinition>>,
+    pub freezable: Option<FreezableRoles<RoleDefinition>>,
+    pub recallable: Option<RecallableRoles<RoleDefinition>>,
+    pub restrict_withdraw: Option<WithdrawableRoles<RoleDefinition>>,
+    pub restrict_deposit: Option<DepositableRoles<RoleDefinition>>,
+}
+
+impl FungibleResourceFeatures {
+    pub fn single_locked_rule(access_rule: AccessRule) -> Self {
+        Self {
+            mintable: mintable! {
+                minter => access_rule.clone(), locked;
+                minter_updater => rule!(deny_all), locked;
+            },
+            burnable: burnable! {
+                burner => access_rule.clone(), locked;
+                burner_updater => rule!(deny_all), locked;
+            },
+            freezable: freezable! {
+                freezer => access_rule.clone(), locked;
+                freezer_updater => rule!(deny_all), locked;
+            },
+            recallable: recallable! {
+                recaller => access_rule.clone(), locked;
+                recaller_updater => rule!(deny_all), locked;
+            },
+            restrict_withdraw: restrict_withdraw! {
+                withdrawer => access_rule.clone(), locked;
+                withdrawer_updater => rule!(deny_all), locked;
+            },
+            restrict_deposit: restrict_deposit! {
+                depositor => access_rule.clone(), locked;
+                depositor_updater => rule!(deny_all), locked;
+            },
+        }
+    }
+
+    pub fn to_features_and_roles(self) -> (Vec<&'static str>, RolesInit) {
+        let mut features = Vec::new();
+        let mut roles = RolesInit::new();
+
+        if let Some(mintable) = &self.mintable {
+            if mintable
+                .minter
+                .ne(&RoleDefinition::locked(AccessRule::DenyAll))
+            {
+                features.push(MINT_FEATURE);
+            }
```

Comment:
Premature optimisation?

Reply:
Yup, this has been fixed in another branch but will just merge that branch into this one